File: Java/FindKPairsWithSmallestSums.java
import java.util.ArrayList;
import java.util.List;
import java.util.PriorityQueue;

public class FindKPairsWithSmallestSums {
    public List<List<Integer>> kSmallestPairs(int[] nums1, int[] nums2, int k) {
        List<List<Integer>> res = new ArrayList<>();
        if (nums1.length == 0 || nums2.length == 0) return res;
        // min-heap of {sum, i, j}
        PriorityQueue<int[]> heap = new PriorityQueue<>((a, b) -> Integer.compare(a[0], b[0]));
        for (int i = 0; i < nums1.length && i < k; i++) {
            heap.add(new int[] { nums1[i] + nums2[0], i, 0 });
        }
        while (!heap.isEmpty() && res.size() < k) {
            int[] cur = heap.poll();
            int i = cur[1], j = cur[2];
            res.add(List.of(nums1[i], nums2[j]));
            if (j + 1 < nums2.length) {
                heap.add(new int[] { nums1[i] + nums2[j + 1], i, j + 1 });
            }
        }
        return res;
    }

    public static void main(String[] args) {
        FindKPairsWithSmallestSums s = new FindKPairsWithSmallestSums();
        System.out.println(s.kSmallestPairs(new int[] { 1, 7, 11 }, new int[] { 2, 4, 6 }, 3)); // [[1, 2], [1, 4], [1, 6]]
        System.out.println(s.kSmallestPairs(new int[] { 1, 1, 2 }, new int[] { 1, 2, 3 }, 2));  // [[1, 1], [1, 1]]
        System.out.println(s.kSmallestPairs(new int[] { 1, 2 }, new int[] { 3 }, 3));           // [[1, 3], [2, 3]]
    }
}

File: Rust/src/bin/find_k_pairs_with_smallest_sums.rs
use std::cmp::Reverse;
use std::collections::BinaryHeap;

fn k_smallest_pairs(nums1: Vec<i32>, nums2: Vec<i32>, k: i32) -> Vec<Vec<i32>> {
    let mut res = Vec::new();
    if nums1.is_empty() || nums2.is_empty() {
        return res;
    }
    let k = k as usize;
    // min-heap of (sum, i, j)
    let mut heap: BinaryHeap<Reverse<(i32, usize, usize)>> = BinaryHeap::new();
    for i in 0..nums1.len().min(k) {
        heap.push(Reverse((nums1[i] + nums2[0], i, 0)));
    }
    while res.len() < k {
        let Reverse((_, i, j)) = match heap.pop() {
            Some(x) => x,
            None => break,
        };
        res.push(vec![nums1[i], nums2[j]]);
        if j + 1 < nums2.len() {
            heap.push(Reverse((nums1[i] + nums2[j + 1], i, j + 1)));
        }
    }
    res
}

fn main() {
    println!("{:?}", k_smallest_pairs(vec![1, 7, 11], vec![2, 4, 6], 3)); // [[1, 2], [1, 4], [1, 6]]
    println!("{:?}", k_smallest_pairs(vec![1, 1, 2], vec![1, 2, 3], 2)); // [[1, 1], [1, 1]]
    println!("{:?}", k_smallest_pairs(vec![1, 2], vec![3], 3)); // [[1, 3], [2, 3]]
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn examples() {
        assert_eq!(
            k_smallest_pairs(vec![1, 7, 11], vec![2, 4, 6], 3),
            vec![vec![1, 2], vec![1, 4], vec![1, 6]]
        );
        assert_eq!(
            k_smallest_pairs(vec![1, 1, 2], vec![1, 2, 3], 2),
            vec![vec![1, 1], vec![1, 1]]
        );
        assert_eq!(
            k_smallest_pairs(vec![1, 2], vec![3], 3),
            vec![vec![1, 3], vec![2, 3]]
        );
    }
}
